### PR TITLE
feat: unlinking db from app functionality

### PR DESCRIPTION
**Pull request by Akirtovskis**

### CODE REVIEW

File: client/src/index.tsx [hunk 1]
```diff
@@ -25,10 +27,17 @@ const GlobalStyle = createGlobalStyle`
   }
 `;
 
-const httpLink = createHttpLink({
+const httpLink = new HttpLink({
   uri: `${config.serverUrl}/graphql`,
 });
 
+const wsLink = new WebSocketLink({
+  uri: `ws://localhost:4000/graphql`,
```
Comment (pradel) [inline client/src/index.tsx:35]: if we harcode the value it won't work when deployed to an external server (on prod)

Comment (pradel) [reply client/src/index.tsx:35]: Maybe we can use `config.serverUrl` but we need to replace the HTTP with ws

File: client/src/index.tsx [hunk 2]
```diff
@@ -53,14 +62,26 @@ const errorLink = onError(({ graphQLErrors }) => {
...
+    return (
+      definition.kind === 'OperationDefinition' &&
+      definition.operation === 'subscription'
+    );
+  },
+  wsLink,
+  httpLink
+);
+
+const client = new ApolloClient({
```
Comment (pradel) [inline client/src/index.tsx:77]: can we keep the name `apolloClient`?

File: server/src/queues/unlinkDatabase.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * - Unlink db
+ */
+const worker = new Worker(
+  queueName,
+  async (job) => {
+    const { app, database } = job.data;
+    debug(
+      `starting unlinkDatabaseQueue for ${database.type} database ${database.name} from  ${app.name} app`
+    );
+    console.log(
```
Comment (pradel) [inline server/src/queues/unlinkDatabase.ts:37]: let's remove all the console.log before merging :)

File: server/src/queues/unlinkDatabase.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    const ssh = await sshConnect();
+
+    const res = await ssh.execCommand(
+      `${dbType}:unlink ${database.name} ${app.name}`
+    );
+
+    const arrayOfLogs = res.stdout.split('\n');
+
+    pubsub.publish('DATABASE_UNLINKED', { unlinkDatabaseLogs: arrayOfLogs });
```
Comment (pradel) [inline server/src/queues/unlinkDatabase.ts:51]: the problem of sending the logs at the end of the exec command is that if the task is taking 10 mins, we will only get the log in the UI 10 mins after. We should rather stream them to have them in real time.

File: server/src/index.ts [hunk 5]
```diff
@@ -219,6 +246,8 @@ const apolloServer = new ApolloServer({
 });
 apolloServer.applyMiddleware({ app });
 
+apolloServer.installSubscriptionHandlers(http);
```
Comment (pradel) [inline server/src/index.ts:249]: Do you think we should remove socket.io in this pr or in another one?

File: server/src/graphql/mutations/unlinkDatabase.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      },
+      include: {
+        apps: {
+          where: {
+            id: appId,
+          },
+        },
+      },
+    }),
+    prisma.app.findOne({
```
Comment (pradel) [inline server/src/graphql/mutations/unlinkDatabase.ts:31]: do you need to do this find as the app is already included in the previous query?

File: server/src/graphql/mutations/unlinkDatabase.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  const isLinked = database.apps.length === 1;
+
+  if (!isLinked) {
+    throw new Error(
+      `${database.name} database is not linked to ${app.name} app`
+    );
+  }
+
+  await unlinkDatabaseQueue.add('unlink-database', {
+    app,
```
Comment (pradel) [inline server/src/graphql/mutations/unlinkDatabase.ts:63]: It's usually better to pass only the id when working with queues (and then do a findOne at the beginning of the queue to retrieve these objects). As the object can be pretty big and exceed the max size allowed.

**Review by pradel — CHANGES_REQUESTED** [review]
First review (mostly server), I will try the UI soon :)

File: client/src/index.tsx (hunk 1 above)
Comment (pradel) [reply client/src/index.tsx:35]: Actually, this won't work `client/src/config.ts`, we need to find a good way to make in work on prod

File: server/src/graphql/mutations/unlinkDatabase.ts (hunk 7 above)
Comment (Akirtovskis) [reply server/src/graphql/mutations/unlinkDatabase.ts]: @pradel fixed.

File: server/src/graphql/mutations/unlinkDatabase.ts (hunk 6 above)
Comment (Akirtovskis) [reply server/src/graphql/mutations/unlinkDatabase.ts]: @pradel no, not really. Fixed!

File: server/src/index.ts (hunk 5 above)
Comment (Akirtovskis) [reply server/src/index.ts:249]: @pradel I am for doing this in a separate PR. Created and issue for this, so we don't forget : https://github.com/ledokku/ledokku/issues/133 .

File: server/src/queues/unlinkDatabase.ts (hunk 4 above)
Comment (Akirtovskis) [reply server/src/queues/unlinkDatabase.ts:60]: @pradel yeah, this one I was not really able to figure out and currently we are sending one bulk of text 😬 .

File: server/src/queues/unlinkDatabase.ts (hunk 3 above)
Comment (Akirtovskis) [reply server/src/queues/unlinkDatabase.ts]: @pradel done.

File: client/src/index.tsx (hunk 1 above)
Comment (Akirtovskis) [reply client/src/index.tsx:35]: Resolving this here as new issue is already opened.

**Review by pradel — APPROVED** [review]
🚀

**Pull request merged**